v2/devdoc/logging.md

```diff
@@ -0,0 +1,162 @@
+
+    logger_log(LOG_LEVEL_VERBOSE, context_1, "log context 1 data");
+    logger_log(LOG_LEVEL_VERBOSE, context_2, "log context 2 data");
+
+    LOG_CONTEXT_DESTROY(context_1);
+
+    // stack allocated context
+    LOCAL_LOG_CONTEXT_DEFINE(local_context, context_2, LOG_CONTEXT_FIELD("the knights that say", "%s", "Nee!"));
+
+    logger_log(LOG_LEVEL_VERBOSE, &local_context, "log with local context");
+
+    LOG_CONTEXT_DESTROY(context_2);
+```
+
+## Log levels
+
+The following log levels shall be exposed by the library:
+
+- `CRITICAL`
+
+- `WARNING`
+
+- `ERROR`
+
+- `INFO`
+
+- `VERBOSE`
+
+## Sinks
+
+Multiple sinks shall be supported. Setting up globally which sinks to be used shall be supported (so that they do not have to be specified every time a `c_logging_log` call is made).
+
+Defining the sinks to use shall be done by defining a NULL terminated array, where each entry shall be a pointer to a structure that represents the sink interface.
+
+Example:
+
+```c
+const LOG_SINK* log_sinks[] = { &console_log_sink, &etw_log_sink, NULL };
+```
+
+### ETW sink
+
+An `ETW` sink shall be supported so that events are emitted via TraceLogging.
+
+For the `ETW` sink there should be a posibility of specifying the provider to use.
+
+### console sink
+
+A console sink shall be supported so that events are printed to the console using `printf`.
+
+## API
+
+```c
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* format, ...);
+```
+
+`log_context` may be `NULL`. If `log_context` is `NULL`, no context information shall be added to the log line.
+
+`format` shall be a `printf`-like format string.
+
+`...` shall be the variable number of arguments required by `format`.
+
+## Contextual logging
+
+For each log line the library shall support specifying a context to be added to the log line.
+
+The context shall allow defining pairs of properties that are added to the event when the event is emitted.
+
+### Local stack context creation
+
+The library shall support defining a context using only stack memory, not requiring any memory allocation:
+
+```c
+#define LOCAL_LOG_CONTEXT_DEFINE(log_context, parent_context, ...) \
+```
+
+where `...` is a list of context fields defined using `LOG_CONTEXT_FIELD`.
+
+Example:
+
+```c
+    LOCAL_LOG_CONTEXT_DEFINE(log_context, NULL, LOG_CONTEXT_FIELD("property_name", "%s", MU_P_OR_NULL(prop_value)));
+```
+
+Bonus:
+
+An ideal implementation for the stack allocated contexts is also to not do any memory copy until actually needed (until they are used).
+
+That would allow wring code like below while not performing the memory copies until the logging is actially executed.
+
+```c
+void a(LOG_CONTEXT_HANDLE log_context, const char* correlation_id)
+{
+    LOCAL_LOG_CONTEXT_DEFINE(local_context, NULL, LOG_CONTEXT_FIELD("correlation_id", "%s", correlation_id));
+
+    if (malloc(42) == NULL)
```

Comment:
>malloc [](http://example.com/codeflow?start=8&length=6)

I fear this might end up with some infinite recursion if logger_log calls malloc internally (and that fails).

Reply:
I'm not sure I got the idea. "a" is an example function (like our bsi_primary_internal_allocate_async.
And my point here was that it would be nice to have the context written once in this function and then used wherever a log error is needed for example.
And ideally it would be awesome if the work of stringifying and memory copy of the bytes if any are needed to only happen if the logging is actually executed.